emma-language/src/main/scala/org/emmalanguage/compiler/lang/backend/TranslateToDataflows.scala

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright © 2014 TU Berlin (emma@dima.tu-berlin.de)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.emmalanguage
+package compiler.lang.backend
+
+import cats.std.all._
+import compiler.Common
+import compiler.lang.core.Core
+import util.Monoids
+
+import shapeless._
+
+/** Translating to dataflows. */
+private[backend] trait TranslateToDataflows extends Common {
+  self: Backend with Core =>
+
+  import UniverseImplicits._
+
+  private[backend] object TranslateToDataflows {
+
+    /**
+     * Translates into a dataflow on the given backend.
+     *
+     * == Preconditions ==
+     *
+     * - The input tree is in Core.
+     *
+     * == Postconditions ==
+     *
+     * - A tree where DataBag operations have been translated to dataflow operations.
+     *
+     * @param backendSymbol The symbol of the target backend.
+     */
+    def translateToDataflows(backendSymbol: u.ModuleSymbol): u.Tree => u.Tree = tree => {
+
+      // Figure out the order stuff
+      val (disambiguatedTree, highFuns): (u.Tree, Set[u.TermSymbol]) = Order.disambiguate(tree)
+
+      val isHighContext: u.Tree =?> Boolean = {
+        case Core.Lang.ValDef(sym, _, _) if highFuns contains sym => true
+      }
```

Comment:
I suggest:

1. Define `isHighContext` as `api.TopDown.inherit(...)` and reuse below.
2. Chain the rest of the transformations with `andThen`.